### PR TITLE
Allow each fully_shard unit to cast foward inputs for mixed precision config

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100290
* #100189

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
sounds good as we discussed the fully_shard and class-based logic differing slightly offline and agreed its fine for now.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -597,18 +605,21 @@ def _root_pre_forward(
...
-        if (
-            should_cast_forward_inputs
-            and state.mixed_precision.cast_root_forward_inputs
-        ):
-            args, kwargs = _cast_forward_inputs(input_dtype, *args, **kwargs)
-        return args, kwargs
+@no_type_check
+def _root_cast_forward_input(state: _FSDPState, args, kwargs) -> Tuple[Any, Any]:
+    should_cast_forward_inputs = (
+        all(not handle._force_full_precision for handle in state._handles)
```
[the remark this answers is not in the document]
Comment (zhaojuanmao) [reply torch/distributed/fsdp/_runtime_utils.py:614]: I intentionally removed this, as 'len(handles) > 0' breaks existing use cases

[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/fsdp/_runtime_utils.py:614]: sounds good

**Pull request closed without merging**